### PR TITLE
ID Bridging PSCCG REVIEW

### PR DESCRIPTION
**Pull request by hillslatt**
ID Bridging changes for Public Comment

### CODE REVIEW

File: implementation.md [hunk 1]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+
+Items in the EID array with <code>mm</code> values of 1 or 2 are generally optional/informational, but some buyers may find it useful for sellers to provide them.
+
+### 7.12.2 - Roles Overview
+- <code>inserter</code> is the party that’s putting the ID into the bid request - Typically a Publisher or an SSP. In other words, the party that made the decision to put the id in the bidstream. In the case of header bidders, the inserter is the Publisher.<br></br>
+- <code>source</code>is the party that defined/created an ID. In the case of universal or alt-IDs, it’s the domain of the party who defined/created the ID itself. In the case of cookie IDs, it’s the domain of the party who the cookie belongs to.<br>
+<b>EXAMPLES FOR PUBLIC COMMENT ONLY</b> adsrvr.org, id5.com, liveramp.com, and others.<br></br>
+- <code>matcher</code> is the party that created the match included in the ID array, this could be a device graph vendor, or the publisher themselves.
+
+### 7.12.3 - Singaling both Agent Type and Match Method
```
Comment (simontrasler) [inline implementation.md:1576]: Typo should be "signaling".

File: 2.6.md [hunk 2]
```diff
@@ -2751,15 +2757,43 @@ Extended identifiers support in the OpenRTB specification allows buyers to use a
     <td><strong>Type&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</strong></td>
     <td><strong>Description</strong></td>
   </tr>
+<tr>
+    <td><code>inserter</code></td>
+    <td>string</td>
+    <td>The canonical domain name of the entity (publisher, publisher monetization company, SSP, Exchange, Header Wrapper, etc) that caused the ID array element to be added. This may be the operational domain of the system, if that is different from the parent corporate domain, to facilitate WHOIS and reverse IP lookups to establish clear ownership of the delegate system.<br>
+    </br>This should be the same value as used to identify sellers in an ads.txt file if one exists.<br>
```
Comment (simontrasler) [inline 2.6.md:2764]: I'd rephrase this slightly, to look more like the following sentence. E.g. (assuming this is a true statement):

> For publishers, this would be the domain used to look up the ads.txt file for the request (if one exists).

File: implementation.md [hunk 3]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+
+## 7.12 - ID Match Method Guidance <a name="idmm"></a>
+
+### 7.12.1 - Best Practices for Disclosing ID Matches
+
+Unless prior arrangements have been made between the buyer and the seller directly, the value in the <code>user.buyeruid</code> field is expected to be derived from a real time cookie sync (see Appendix: Cookie Syncing) and value in <code>device.ifa</code> field is expected to be derived from an advertising ID call to the Operating System. 
+
+For an Exchange to properly disclose that an ID substitution has occurred by a publisher, they must propagate the <code>mm</code> value, as sent by the publisher, on the Extended Identifier (eid) they send to the DSP, regardless of how the SSP determined the DSP’s id.
```
Comment (simontrasler) [inline implementation.md:1564]: Capitalize `ID` everywhere.

File: implementation.md [hunk 4]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+### 7.12.4 - json Examples
+
+#### 7.12.4.1 User and Device IDs
+
+##### Example 1. DSP cookie derived from real-time cookie sync, subsequently observable
+```
+{
+    "user": {
+        "buyeruid": "fac13741-0648-436a-88cf-aceafdf45c9a",
+        "eids": [
```
Comment (simontrasler) [inline implementation.md:1613]: This usage of `eids` is weird. What if there are other EIDs in the array?

File: 2.6.md [hunk 5]
```diff
@@ -2751,15 +2757,43 @@ Extended identifiers support in the OpenRTB specification allows buyers to use a
...
+    <td><code>mm</code></td>
+    <td>int</td>
+    <td>Match method used by the <code>matcher</code>. Refer to <a href="https://github.com/InteractiveAdvertisingBureau/AdCOM/blob/master/AdCOM%20v1.0%20FINAL.md#list_agenttypes">List: ID Match Methods</a> in AdCOM 1.0
+    </td>
   </tr>
   <tr>
     <td><code>uids</code></td>
     <td>object array</td>
-    <td>Array of extended ID <code>UID</code> objects from the given source. Refer to 3.2.28 Extended Identifier UIDs</td>
+    <td>Array of extended ID <code>UID</code> objects from the given source. Refer to <a href=https://github.com/InteractiveAdvertisingBureau/openrtb2.x/blob/main/2.6.md#3228---object-uid-"3.2.28"</a> Extended Identifier UIDs</td>
```
Comment (robhazan) [inline 2.6.md:2796]: This URL is broken, and we shouldn't use HTML anchors or hard links of this form in here anyway. If we want to link, let's use a relative link like this `[Extended Identifier UIDs](#objectuid)`.  Though everywhere else we refer to other sections in this document, we seem to just refer to the section (e.g. `<code>Extended Identifier UIDs</code> object (Section 3.2.28)`).

File: implementation.md (hunk 3 above)
Comment (robhazan) [inline implementation.md:1564]: I think you can get rid of this 1-sentence paragraph, because the following paragraph describes this exact same scenario.

File: implementation.md [hunk 6]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+- <code>source</code>is the party that defined/created an ID. In the case of universal or alt-IDs, it’s the domain of the party who defined/created the ID itself. In the case of cookie IDs, it’s the domain of the party who the cookie belongs to.<br>
+<b>EXAMPLES FOR PUBLIC COMMENT ONLY</b> adsrvr.org, id5.com, liveramp.com, and others.<br></br>
+- <code>matcher</code> is the party that created the match included in the ID array, this could be a device graph vendor, or the publisher themselves.
+
+### 7.12.3 - Singaling both Agent Type and Match Method
+Values in the <code>atype</code> attribute should be used to describe the level of the ID in the bidstream and can apply to multiple match methods. 
+
+Different iterations of possible combinations are listed in the following table. Additional json examples for how to use <code>atype</code> alongside match method (<code>mm</code>) signaling can be found in the next section.
+
+<table>
```
Comment (robhazan) [inline implementation.md:1581]: Minor nit -- all instances of "mm" in the below should use code (i.e. ``mm``).

File: implementation.md (hunk 4 above)
Comment (robhazan) [reply implementation.md:1613]: Yeah, this should not be here. atype=1 is inferred from the existence of buyeruid.

File: implementation.md [hunk 7]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+                }
+            ]
+        },
+        {
+            "inserter": "sspdomain.com",
+            "source": "dspdomain.com”,
+            "mm": 3,
+            "uids": [
+                {
+                    "id": "3d89748d-74bb-44f1-9908-298090dc2944"
```
Comment (robhazan) [inline implementation.md:1700]: why no "atype" here?

File: implementation.md [hunk 8]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+"user": {
+    "buyeruid": "fac13741-0648-436a-88cf-aceafdf45c9a",
+    "eids": [
+        {
+            "source": "sspdomain.com”,
+            "mm": 1,
+            "uids": [
+                {
+                    "id": "3d89748d-74bb-44f1-9908-298090dc2944"
+		    "atype": "3"
```
Comment (robhazan) [inline implementation.md:1723]: this should be "atype":1 -- it's the SSP's own cookie from the browser.

File: implementation.md [hunk 9]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+            ]
+        },
+        {
+            "matcher": "idvendor.com",
+	    "inserter": "sspdomain.com",
+            "source": "universalid.com”,
+            "mm": 4,
+            "uids": [
+                {
+                    "id": "fac13741-0648-436a-88cf-aceafdf45c9a"
```
Comment (robhazan) [inline implementation.md:1734]: this is missing "atype", and I believe it should be "atype":3

File: implementation.md [hunk 10]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+        "eids": [
+            {
+            "inserter": "pub.com",
+            "matcher": "graphvendor.com",
+            "source": "sspdomain.com”,
+            "mm": "3”,
+                "uids": [
+                    {
+                        "id": "fac13741-0648-436a-88cf-aceafdf45c9a",
+			"atype": "1"
```
Comment (robhazan) [inline implementation.md:1661]: shouldn't this be "atype":3, since it's a hashed email, which is the same across devices?

File: implementation.md [hunk 11]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+    "eids": [
+        {
+            "matcher": "graphvendor.com",
+            "inserter": "pub.com",
+            "source": "sspdomain.com”,
+            "mm": 3,
+            "uids": [
+                {
+                    "id": "fac13741-0648-436a-88cf-aceafdf45c9a"
+		    "atype": "1"
```
Comment (robhazan) [inline implementation.md:1690]: again, this should be "atype":3, I believe.

File: implementation.md (hunk 7 above)
Comment (robhazan) [reply implementation.md:1700]: And I think it should be "atype":1 (which is a little funky because it's based on a hashed email, which is itself "atype":3), because the DSP's cookie is associated with 1 specific browser. Someone could reasonably argue that `atype` should inherit "the weakest link", just like `mm` does...

File: implementation.md [hunk 12]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+            ]
+        },
+        {
+            "matcher": "app-pub.com",
+	    "inserter": "app-pub.com",
+            "source": "alt-id.org”,
+            "mm": 3,
+            "uids": [
+                {
+                    "id": "e7f8c5f2-6b4e-4565-8b0a-53d8b730143e"
```
Comment (robhazan) [inline implementation.md:1767]: missing "atype":3

Comment (robhazan) [reply implementation.md:1767]: Though I suppose depending on the ID (if it's partitioned by app bundle), could technically be "atype":2, as well.

File: implementation.md [hunk 13]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+    "buyeruid": "3d89748d-74bb-44f1-9908-298090dc2944",
+    "eids": [
+        {
+            "matcher": "graphvendor.com",
+            "inserter": "pub.com",
+            "source": "sspdomain.com”,
+            "mm": 3,
+            "uids": [
+                {
+                    "id": "fac13741-0648-436a-88cf-aceafdf45c9a"
```
Comment (robhazan) [inline implementation.md:1689]: missing a comma on the end

File: implementation.md [hunk 14]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+"user": {
+    "eids": [
+        {
+	    "matcher": "app-pub.com",
+	    "inserter": "app-pub.com",
+	    "source": "apple.com”,
+            "mm": 3,
+            "uids": [
+                {
+                    "id": "fac13741-0648-436a-88cf-aceafdf45c9a"
```
Comment (robhazan) [inline implementation.md:1755]: missing a comma on the end

File: implementation.md [hunk 15]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+"user": {
+    "eids": [
+        {
+	    "matcher": "graphvendor.com",
+	    "inserter": "app-pub.com",
+	    "source": "apple.com”,
+            "mm": 4,
+            "uids": [
+                {
+                    "id": "fac13741-0648-436a-88cf-aceafdf45c9a"
```
Comment (robhazan) [inline implementation.md:1791]: missing a comma on the end

File: implementation.md [hunk 16]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+		    "atype": "3"
```
Comment (robhazan) [inline implementation.md:1792]: shouldn't this be "atype":2?

File: implementation.md [hunk 17]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+{
+"user": {
+    "eids": [
+        {
+	    "inserter": "app-pub.com",
+	    "source": "app-pub.com”,
+            "mm": "6",
+            "uids": [
+                {
+                    "id": "fac13741-0648-436a-88cf-aceafdf45c9a"
```
Comment (robhazan) [inline implementation.md:1813]: missing a comma on the end

File: implementation.md [hunk 18]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+    }
+}
+```
+
+##### Example 2. Resettable advertising ID generated by operating system
+```
+{
+    "device": {
+        "ifa": "df3a42f4-a388-43e9-a691-b0d308c963b4",
+        "eids": [
```
[the remark this answers is not in the document]
Comment (simontrasler) [reply implementation.md:1631]: And there should be no scenario of an `eids` array that does not contain IDs, correct?

File: implementation.md (hunk 3 above)
Comment (hillslatt) [reply implementation.md:1564]: i think the more times we repeat it the better

File: 2.6.md (hunk 5 above)
Comment (hillslatt) [reply 2.6.md:2796]: fixed...i think

File: 2.6.md (hunk 2 above)
Comment (hillslatt) [reply 2.6.md:2764]: fixed per our conversation via email

File: implementation.md (hunk 1 above)
Comment (hillslatt) [reply implementation.md:1576]: thank you! fixed.

File: implementation.md (hunk 3 above)
Comment (hillslatt) [reply implementation.md:1564]: done

File: implementation.md (hunk 4 above)
Comment (hillslatt) [reply implementation.md:1613]: i get very nervous with any kind of inference when it comes to this stuff. also i wanted to include atype in all the examples, even where they may be redundant to give the community a visual for how to use them in conjunction with each other. not a hill i'll die on, but do we really feel strongly we should nuke it?

File: implementation.md [hunk 19]
```diff
@@ -1554,3 +1554,268 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+
+Unless prior arrangements have been made between the buyer and the seller directly, the value in the <code>user.buyeruid</code> field is expected to be derived from a real time cookie sync (see Appendix: Cookie Syncing) and value in <code>device.ifa</code> field is expected to be derived from an advertising ID call to the Operating System. 
+
+For an Exchange to properly disclose that an ID substitution has occurred by a publisher, they must propagate the <code>mm</code> value, as sent by the publisher, on the Extended Identifier (eid) they send to the DSP, regardless of how the SSP determined the DSP’s id.
+
+If a linkage is made by any party (e.g. a SSP to a DSP cookie ID) based on an incoming EID where <code>mm</code> does not equal 1 or 2, then the ID sent to the recipient must inherit the <code>mm</code> value of the incoming EID. For example, if a SSP receives an ID based on any kind of matching (<code>mm</code> values 3,4,5,6), then even if the ID they send to the DSP is based on a cookie exchange, it must also indicate that matching occurred (i.e. <code>mm</code> values 3,4,5,6). See example #4 below as an illustration.
```
[the remark this answers is not in the document]
Comment (hillslatt) [reply implementation.md:1566]: updated per our slack convo

File: implementation.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (hillslatt) [reply implementation.md:1631]: this is just to show the atype in there. we can nuke it after public comment is done but i want as many examples of both in there as possible. did i get the code wrong?

File: implementation.md (hunk 7 above)
Comment (hillslatt) [reply implementation.md:1700]: added atype to the ssp node. i'll let @patmmccann weigh in on which atype value it should be. its' 1 for now

File: implementation.md (hunk 8 above)
Comment (hillslatt) [reply implementation.md:1723]: but its a person based because its across devices no?

Comment (hillslatt) [reply implementation.md:1723]: wait...no...i fixed. its 1 in the first node and 3 in the second...i think. please double check me

File: implementation.md (hunk 9 above)
Comment (hillslatt) [reply implementation.md:1734]: added

File: implementation.md (hunk 11 above)
Comment (hillslatt) [reply implementation.md:1690]: done

File: implementation.md (hunk 12 above)
Comment (hillslatt) [reply implementation.md:1767]: added as 3, but i'll defer to you and @patmmccann for all the atype values

File: implementation.md (hunk 4 above)
Comment (adammarkey) [reply implementation.md:1614]: By definition, I think we can infer everything on buyeruid. Example description should be updated noting this. We can fix after public comment IMO

Comment (robhazan) [reply implementation.md:1614]: Agreed. The empty EIDs array makes no sense to me. I think simply remove. No need for atype.

File: implementation.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (robhazan) [reply implementation.md:1632]: Here too, I think we should remove. Don't show how to do things that we don't want implementers doing. We don't want implementers adding empty eids arrays with just `atype` in them for buyeruid or ifa.

File: implementation.md [hunk 20]
```diff
@@ -1554,3 +1554,272 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+    "eids": [
+        {
+	    "matcher": "graphvendor.com",
+	    "inserter": "app-pub.com",
+	    "source": "apple.com”,
+            "mm": 4,
+            "uids": [
+                {
+                    "id": "fac13741-0648-436a-88cf-aceafdf45c9a",
+		    "atype": "3",
```
Comment (robhazan) [inline implementation.md:1796]: shouldn't this be `"atype": 2` rather than 3? An Apple IFA is a device ID, even if the match-method is based on inference.

File: implementation.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (robhazan) [reply implementation.md:1632]: Sidenote, @hillslatt : `atype` is an integer, not a string, so everywhere you provided examples of it, it should be (for example) `"atype": 1`, not `"atype": "1"`.

File: implementation.md [hunk 21]
```diff
@@ -1554,3 +1554,272 @@ In the `Video` and `Audio` objects, sellers must only include one form of floor
...
+"user": {
+    "eids": [
+        {
+	    "inserter": "app-pub.com",
+	    "source": "app-pub.com”,
+            "mm": "6",
+            "uids": [
+                {
+                    "id": "fac13741-0648-436a-88cf-aceafdf45c9a",
+		    "atype": "2",
```
Comment (robhazan) [inline implementation.md:1818]: Pretty sure this shouldn't be `atype` 1 or 2 or 3. It doesn't match any of the definitions. Do we need to add a new atype for a session ID? `"atype": 4`?

File: implementation.md (hunk 20 above)
Comment (hillslatt) [reply implementation.md:1796]: i had it as 3 because its x-device so it'd be a person based identifier

Comment (robhazan) [reply implementation.md]: I don't think that's the spirit of an "agent type". From AdCOM: "This list identifies the user agent types a user identifier is from." An IFA from an Apple device is a device identifier, not a user identifier. This is why I made the callout above that it's still a device ID, even if it was matched in based on inference that it's the same user at that IP. It has not become a RampID or something -- it's still an IFA.
